api/genericresource/resource_management.go

```diff
@@ -0,0 +1,62 @@
+package genericresource
+
+import (
+	"github.com/docker/swarmkit/api"
+)
+
+// Claim assigns GenericResources to the task by taking them from the node's GenericResource
+// and storing them in the taken store
+func Claim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
+	t, err := getGenericResourceDiscrete(taskRes)
+	if err != nil {
+		return err
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.GenericResource_Discrete:
+		if nr.Discrete < t.Discrete {
+			return NewError("cannot provide enough of scalar resource requested %d, available %d",
+				t.Discrete, nr.Discrete)
+		}
+		nr.Discrete -= t.Discrete
+		taken.Resource = NewDiscrete(t.Discrete)
+	case *api.GenericResource_Set:
+		i := uint64(0)
+		res := NewSet()
+
+		for k := range nr.Set.Val {
+			if i >= t.Discrete {
+				break
+			}
+
+			i++
+
+			res.Set.Val[k] = nil
+			if _, ok := nr.Set.Val[k]; !ok {
+				return NewError("missing key %s from node resource", k)
+			}
+			delete(nr.Set.Val, k)
+		}
+
+		taken.Resource = res
+	}
+
+	return nil
+}
+
+// Reclaim adds the resources taken by the task to the taken store
+func Reclaim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
```

Comment:
I think `nodeRes` can end up being nil here if the node was updated to no longer have that type of resource. Either `Reclaim` should check for this, or the scheduler should avoid calling `Reclaim` in this case.